NorthstarDedicatedTest/AntiRCE.cpp

```diff
@@ -0,0 +1,248 @@
+// Because this could be called by all sorts of external software and windows procedures, the only check here is for writing to binaries
+// Windows Documentation: https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-zwcreatefile
+void* oNtCreateFile;
+LONG WINAPI hkNtCreateFile(
+	PHANDLE pFileHandle,
+	ACCESS_MASK desiredAccess,
+	PVOID pObjectAttributes,
+	PVOID pIoStatusBlock,
+	PLARGE_INTEGER pAllocationSize,
+	ULONG fileAttributes,
+	ULONG shareAccess,
+	ULONG createDisposition,
+	ULONG createOptions,
+	PVOID eaBuffer,
+	ULONG eaLength)
+{
+
+	// Call original first, see if this call even succeeded
+	auto ofunc = (decltype(&hkNtCreateFile))oNtCreateFile;
+	LONG result = ofunc(
+		pFileHandle,
+		desiredAccess,
+		pObjectAttributes,
+		pIoStatusBlock,
+		pAllocationSize,
+		fileAttributes,
+		shareAccess,
+		createDisposition,
+		createOptions,
+		eaBuffer,
+		eaLength);
+
+	// clang-format off
+	// All ACCESS_MASK flags that allow modification of the file
+	constexpr DWORD ALL_WRITE_FLAGS[] = {
+		GENERIC_WRITE, GENERIC_ALL, 
+		WRITE_DAC, WRITE_OWNER, 
+		FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_WRITE_EA, FILE_WRITE_ATTRIBUTES,
+		FILE_DELETE_CHILD, DELETE
+	};
+	// clang-format on
+
+	bool hasWriteAccess = false;
+	for (DWORD writeFlag : ALL_WRITE_FLAGS)
+	{
+		if (desiredAccess & writeFlag)
+		{
+			hasWriteAccess = true;
+			break;
+		}
+	}
+
+	if (hasWriteAccess)
+	{
+		char filePathBuf[MAX_PATH] = {};
+		if (GetFinalPathNameByHandleA(*pFileHandle, filePathBuf, MAX_PATH, FILE_NAME_NORMALIZED))
+		{
+			const char* extension = strrchr(filePathBuf, '.');
+			if (extension)
+			{
+				for (auto evil : EVIL_EXTENTIONS)
+				{
+					if (!_stricmp(evil, extension))
+					{
+						AntiRCE::EmergencyReport("Tried to access file \"" + std::string(filePathBuf) + "\"");
+					}
+				}
+			}
+		}
+	}
+
+	return result;
+}
+
+// NOTE: These hooks do not prevent RCE attacks from using these funtions, as an attacker could just disable these hooks
+// (particularly easy if you get a ROP-chain going and use MH_RemoveHook lol)
+//
+// These hooks are only meant to be a barrier of entry for RCEs,
+//	as well as to prevent any exploits that might utilize these functions from being able to do actual damage
+void AntiRCE::EmergencyReport(std::string msg)
+{
+#ifdef _DEBUG // For easier debugging
+	assert(false, "AntiRCE::EmergencyReport triggered");
+#endif
+
+	spdlog::critical("AntiRCE EMERGENCY REPORT: " + msg);
+
+#ifndef _DEBUG // Disabled in debugging, gets annoying during testing
+
+	// Beep an out-of-tune tri-tone for 300ms so user knows something is up
+	Beep(494 * 2, 150);
+	Beep(349 * 2, 150);
+
+#endif
+
+	// clang-format off
+	MessageBoxA(NULL, (
+			"AntiRCE EMERGENCY SHUTDOWN:\n" + msg + 
+			"\n\nIt is HIGHLY reccommended that you notify northstar developers/staff of this issue (unless you are sure it's fine)."
+		).c_str(), 
+		"Northstar AntiRCE", 
+		MB_ICONERROR
+	);
```

Comment:
I get that it is important that we get informed of an e.g. out-of-bounds write but I'm not sure this is the best approach. A crash and remote logger like [Sentry.io](https://sentry.io/) might be a better solution but we don't have support for that yet.

For now, I'd just drop the beeping noise.

Reply:
i fix 👍